flink-core/src/main/java/org/apache/flink/metrics/groups/TaskManagerJobMetricGroup.java

```diff
@@ -0,0 +1,122 @@
+ * limitations under the License.
+ */
+package org.apache.flink.metrics.groups;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricRegistry;
+import org.apache.flink.metrics.groups.scope.ScopeFormat.TaskManagerJobScopeFormat;
+import org.apache.flink.util.AbstractID;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Special {@link org.apache.flink.metrics.MetricGroup} representing everything belonging to
+ * a specific job, running on the TaskManager.
+ *
+ * <p>Contains extra logic for adding Tasks ({@link TaskMetricGroup}).
+ */
+@Internal
+public class TaskManagerJobMetricGroup extends JobMetricGroup {
+
+	/** The metrics group that contains this group */
+	private final TaskManagerMetricGroup parent;
+
+	/** Map from execution attempt ID (task identifier) to task metrics */
+	private final Map<AbstractID, TaskMetricGroup> tasks = new HashMap<>();
+
+	// ------------------------------------------------------------------------
+
+	public TaskManagerJobMetricGroup(
+		MetricRegistry registry,
+		TaskManagerMetricGroup parent,
+		JobID jobId,
+		@Nullable String jobName) {
+
+		this(registry, checkNotNull(parent), registry.getScopeFormats().getTaskManagerJobFormat(), jobId, jobName);
+	}
+
+	public TaskManagerJobMetricGroup(
+		MetricRegistry registry,
+		TaskManagerMetricGroup parent,
+		TaskManagerJobScopeFormat scopeFormat,
+		JobID jobId,
+		@Nullable String jobName) {
+
+		super(registry, jobId, jobName, scopeFormat.formatScope(parent, jobId, jobName));
+
+		this.parent = checkNotNull(parent);
+	}
+
+	public final TaskManagerMetricGroup parent() {
+		return parent;
+	}
+
+	// ------------------------------------------------------------------------
+	//  adding / removing tasks
+	// ------------------------------------------------------------------------
+
+	public TaskMetricGroup addTask(
+		AbstractID vertexId,
+		AbstractID executionId,
+		String taskName,
+		int subtaskIndex,
+		int attemptNumber) {
+
+		checkNotNull(executionId);
+
+		synchronized (this) {
+			if (!isClosed()) {
+				TaskMetricGroup task = new TaskMetricGroup(registry, this,
+					vertexId, executionId, taskName, subtaskIndex, attemptNumber);
+				tasks.put(executionId, task);
+				return task;
+			} else {
+				return null;
+			}
+		}
+	}
+
+	public void removeTaskMetricGroup(AbstractID executionId) {
+		checkNotNull(executionId);
+
+		boolean removeFromParent = false;
+		synchronized (this) {
+			if (!isClosed() && tasks.remove(executionId) != null && tasks.isEmpty()) {
+				// this call removed the last task. close this group.
+				removeFromParent = true;
+				close();
+			}
+		}
+
+		// IMPORTANT: removing from the parent must happen while holding the this group's lock,
+		//      because it would violate the "first parent then subgroup" lock acquisition order
```

Comment:
I would assume that it is a typo. The `TaskManagerMetricGroup.removeJobMetricsGroup` does not access the `TaskManagerJobMetricGroup` where it could try to acquire the lock for the task manager job metric group.